### PR TITLE
Qospolicy controller and host-agent side changes

### PR DESCRIPTION
**Pull request by niteshcha**

### CODE REVIEW

**Comment by niteshcha** [comment]
@tomflynn @snaiksat please review this!

**Comment by jayaramsatya** [comment]
1. Please fix the test case failure  it is crashing,  you need to add fake controller initialization for your CRD.
controller/common_test.go please refer this file you can find sample code.
2. Also run the go fmt to fix the indentation

**Comment by niteshcha** [comment]
I have added the UT and fake controller as well.
Indentation has been fixed.
Validation is there in our CRD schema for policing rate and policing burst...
we have defined them as integer type.


please review the updated change set.

File: pkg/hostagent/pods.go [hunk 1]
```diff
@@ -464,21 +501,26 @@ func (agent *HostAgent) podChangedLocked(podobj interface{}) {
 	if epAttributes == nil {
 		epAttributes = make(map[string]string)
 	}
+	qosPolicies, _ := agent.assignqosPolicies(pod)
 	epAttributes["vm-name"] = pod.ObjectMeta.Name
 	epAttributes["namespace"] = pod.ObjectMeta.Namespace
 
-	agent.epChanged(&epUuid, &epMetaKey, &epGroup, secGroup, epAttributes, logger)
+	logger.WithFields(logrus.Fields{
```
Comment (jayaramsatya) [inline pkg/hostagent/pods.go:508]: Please remove this logs here

File: pkg/hostagent/pods.go [hunk 2]
```diff
@@ -443,6 +446,40 @@ func (agent *HostAgent) podChanged(podkey *string) {
 	agent.podChangedLocked(podobj)
 }
 
+func addPolicy(gset map[metadata.OpflexGroup]bool, g []metadata.OpflexGroup,
```
Comment (jayaramsatya) [inline pkg/hostagent/pods.go:449]: seems this code is repeated similar to addGroup.  you can move both these functions to group_assign.go

File: pkg/hostagent/pod_relatives.go [hunk 3]
```diff
@@ -181,6 +183,90 @@ func (agent *HostAgent) initNetPolPodIndex() {
...
+}
+
+func (agent *HostAgent) qosPolicyChanged(oldobj, newobj interface{}) {
+	oldnp := oldobj.(*qospolicy.QosPolicy)
+	newnp := newobj.(*qospolicy.QosPolicy)
+	if !reflect.DeepEqual(oldnp.Spec.Selector, newnp.Spec.Selector) {
+		agent.qosPolPods.UpdateSelectorObjNoCallback(newobj)
+	}
+
+	npkey, err := cache.MetaNamespaceKeyFunc(newnp)
```
Comment (jayaramsatya) [inline pkg/hostagent/pod_relatives.go:226]: NIt:  can you change this name to npkey to qpkey

File: pkg/qospolicy/apis/aci.qos/v1/types.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+// QosPolicyList contains a list of QosPolicy
+type QosPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []QosPolicy `json:"items"`
+}
+
+type PodSelector struct {
+	Labels    map[string]string `json:"labels,omitempty"`
+	Namespace string            `json:"namespace,omitempty"`
```
Comment (jayaramsatya) [inline pkg/qospolicy/apis/aci.qos/v1/types.go:65]: So here assumption is  Qos Policy will be applied at namespace or Pod level ?

File: pkg/qospolicy/apis/aci.qos/v1/types.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	PolicingRate  int `json:"policing_rate"`
+	PolicingBurst int `json:"policing_burst"`
+}
+
+// QosPolicyStatus defines the observed state of QosPolicy
+type QosPolicyStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file7
+	// Add custom validation using kubebuilder tags: https://book.kubebuilder.io/beyond_basics/generating_crd.html
+	State PolicyState `json:"state"`
```
Comment (jayaramsatya) [inline pkg/qospolicy/apis/aci.qos/v1/types.go:38]: Where are you setting your policy state in code? are you planning to use it for future ?

File: pkg/hostagent/pods.go (hunk 1 above)
Comment (niteshcha) [reply pkg/hostagent/pods.go]: done

File: pkg/hostagent/pods.go (hunk 2 above)
Comment (niteshcha) [reply pkg/hostagent/pods.go]: done

File: pkg/hostagent/pod_relatives.go (hunk 3 above)
Comment (niteshcha) [reply pkg/hostagent/pod_relatives.go]: done

File: pkg/qospolicy/apis/aci.qos/v1/types.go (hunk 4 above)
Comment (niteshcha) [reply pkg/qospolicy/apis/aci.qos/v1/types.go:65]: Qos Policy will be applied at namespace level.Pod level also we will address with EP annotation in separate commit

File: pkg/qospolicy/apis/aci.qos/v1/types.go (hunk 5 above)
Comment (niteshcha) [reply pkg/qospolicy/apis/aci.qos/v1/types.go:38]: yes...planning to use in future

**Comment by jayaramsatya** [comment]
LGTM, Please wait merging the code until acc-provision changes are  in. otherwise cluster rollback issues will popup

**Comment by jayaramsatya** [comment]
LGTM, could you please check it requires any rebase ?

**Comment by niteshcha** [comment]
> LGTM, could you please check it requires any rebase ?

yes it was required and it's done...

**Comment by niteshcha** [comment]
> LGTM, could you please check it requires any rebase ?

rebase done

**Pull request merged**